src/table.js

```diff
@@ -542,15 +551,79 @@ export function getTypeValidator(colType) {
   }
 }
 
+export function coerceToType(value, type) {
+  switch (type) {
+    case "string":
+      return typeof value === "string" || value == null
+        ? value
+        : String(value);
+    case "boolean":
+      if (typeof value === "string") {
+        const trimValue = value.trim();
+        return trimValue === "true" ? true : trimValue === "false" ? false : null;
+      }
+      return typeof value === "boolean" || value == null
+        ? value
+        : Boolean(value);
+    case "bigint":
+      return typeof value === "bigint" || value == null
+        ? value
+        : Number.isInteger(typeof value === "string" && !value ? NaN : +value)
+        ? BigInt(value) // eslint-disable-line no-undef
+        : undefined;
+    case "integer": // not a target type for coercion, but can be inferred
+    case "number": {
+      return typeof value === "number"
+        ? value
+        : value == null || (typeof value === "string" && !value)
+        ? NaN
+        : Number(value);
+    }
+    case "date": {
+      if (value instanceof Date) return value;
+      const trimValue = typeof value === "string" ? value.trim() : value;
+      return /^(([-+]\d{2})?\d{4}(-\d{2}(-\d{2}))|(\d{1,2})\/(\d{1,2})\/(\d{2,4}))([T ]\d{2}:\d{2}(:\d{2}(\.\d{3})?)?(Z|[-+]\d{2}:\d{2})?)?$/.test(trimValue)
+        ? new Date(trimValue)
+        : new Date("");
```

Comment:
Assuming we handle null, undefined, and numbers as above, I think at this point it makes sense to explicitly coerce the value to a string. We’re doing that _implicitly_ when we test against the regex, but I think it’d be better to do it _explicitly_ like so:

```suggestion
      const trimValue = String(value).trim();
      return new Date(
        /^(([-+]\d{2})?\d{4}(-\d{2}(-\d{2}))|(\d{1,2})\/(\d{1,2})\/(\d{2,4}))([T ]\d{2}:\d{2}(:\d{2}(\.\d{3})?)?(Z|[-+]\d{2}:\d{2})?)?$/.test(
          trimValue
        )
          ? trimValue
          : NaN
      );
```

Alternatively if we want to return undefined here instead of invalid dates:

```suggestion
      const trimValue = String(value).trim();
      return /^(([-+]\d{2})?\d{4}(-\d{2}(-\d{2}))|(\d{1,2})\/(\d{1,2})\/(\d{2,4}))([T ]\d{2}:\d{2}(:\d{2}(\.\d{3})?)?(Z|[-+]\d{2}:\d{2})?)?$/.test(
        trimValue
      )
        ? new Date(trimValue)
        : undefined;
```

Reply:
Chiming in briefly - (hope it's helpful!) - IMO, if a value isn't coercible to a date, we should return an invalid Date object to signal to users that a row isn't in the expected format. If a user saw `undefined`, they may think that the row is empty, rather than improperly formatted. 

Also, sorry if I missed this in an earlier conversation, but couldn't we just return `new Date(value)` in any case that `value` is defined? Then the `Date` constructor will handle returning `invalid Date`, and will also coerce numbers to dates (which feels appropriate). I briefly tested with:

```js
// Date constructor returns invalid date for incorrect formats:
new Date("what")
new Date([])
new Date({})
new Date(new ArrayBuffer)

// This is maybe expected, treating `true` as 1
new Date(true) // Wed Dec 31 1969 19:00:00 GMT-0500 (Eastern Standard Time)

// This I find surprising - that 0 and 1 return the same time 🤔 
new Date(false) // Wed Dec 31 1969 19:00:00 GMT-0500 (Eastern Standard Time)
```

Reply:
The reason we don’t want to “`new Date(value)` in any case” is mentioned in https://github.com/observablehq/stdlib/pull/346/files/6bee09b37b3b0a32213203699a82bf306d7e17ae#diff-6e1c3dc6cd2f88ca61d65a2f261fd5adfb15f103d719af5acaa064433d3b7bee: 

> The behavior of the Date constructor is not consistent across browsers; some browsers (Chrome) are much more lenient than others (Safari). Maybe it’s okay to be as lenient as possible here, since we do a regex test when computing the inferred types for strings… but potentially we could repeat that regex test here, or perhaps use [isoformat’s parse](https://github.com/mbostock/isoformat#parsedate-fallback) (which we already use elsewhere) to only allow ISO 8601 formatted dates to be coerced.

I like the principle of distinguishing between missing and invalid/non-coercible values. That suggests preserving nullish values when coercing to dates, coercing the empty string to null, and coercing non-ISO 8601 date strings to “Invalid Date” objects.

Reply:
Thanks for clarifying!